lib/Dialect/Arc/Transforms/LowerClocksToFuncs.cpp

```diff
@@ -0,0 +1,191 @@
+                          << "`\n");
+
+  // Create a call to the function within the model.
+  builder.setInsertionPoint(clockOp);
+  builder.create<func::CallOp>(clockOp->getLoc(), funcOp,
+                               ValueRange{modelStorageArg});
+
+  // Move the clock's body block to the function and remove the old clock op.
+  funcOp.getBody().takeBody(clockRegion);
+  clockOp->erase();
+
+  return success();
+}
+
+/// Copy any external constants that the clock tree might be using into its
+/// body. Anything besides constants should no longer exist after a proper run
+/// of the pipeline.
+LogicalResult LowerClocksToFuncsPass::isolateClock(Operation *clockOp,
+                                                   Value modelStorageArg,
+                                                   Value clockStorageArg) {
+  auto *clockRegion = &clockOp->getRegion(0);
+  auto builder = OpBuilder::atBlockBegin(&clockRegion->front());
+  DenseMap<Value, Value> copiedValues;
+  auto result = clockRegion->walk([&](Operation *op) {
+    for (auto &operand : op->getOpOperands()) {
+      // Block arguments are okay, since there's nothing we can move.
+      if (operand.get() == modelStorageArg) {
+        operand.set(clockStorageArg);
+        continue;
+      }
+      if (operand.get().isa<BlockArgument>())
+        continue;
+
+      // Check if the value is defined outside of the clock op.
+      auto *definingOp = operand.get().getDefiningOp();
+      assert(definingOp && "block arguments ruled out above");
+      Region *definingRegion = definingOp->getParentRegion();
+      if (clockRegion->isAncestor(definingRegion))
+        continue;
+
+      // The op is defined outside the clock, so we need to create a copy of the
+      // defining inside the clock tree.
+      if (auto copiedValue = copiedValues.lookup(operand.get())) {
+        operand.set(copiedValue);
+        continue;
+      }
+
+      // Check that we can actually copy this definition inside.
+      if (!definingOp->hasTrait<ConstantLike>()) {
+        auto d = op->emitError("operation in clock tree uses external value");
+        d.attachNote() << "clock trees can only use external constant values";
+        d.attachNote(definingOp->getLoc()) << "external value defined here:";
+        d.attachNote(clockOp->getLoc()) << "clock tree:";
+        return WalkResult::interrupt();
+      }
+
+      // Copy the op inside the clock tree (or move it if all uses are within
+      // the clock tree).
+      bool canMove = llvm::all_of(definingOp->getUsers(), [&](Operation *user) {
+        return clockRegion->isAncestor(user->getParentRegion());
+      });
+      Operation *clonedOp;
+      if (canMove) {
+        definingOp->remove();
+        clonedOp = definingOp;
+        ++numOpsMoved;
+      } else {
+        clonedOp = definingOp->cloneWithoutRegions();
+        ++numOpsCopied;
+      }
+      builder.setInsertionPoint(op);
```

Comment:
Choosing this insertion point combined with caching the clone in `copiedValue` results in problems with nested regions (e..g., when it is used in two separate nested regions).
As we only clone constants anyways, we should insert them at the start of the clock region.

Modifying the regression test to the following exposes the issue:
```
arc.model "Trivial" {
^bb0(%arg0: !arc.storage<42>):
  %true = hw.constant true
  %false = hw.constant false
  arc.clock_tree %true {
    %c0_i9001 = hw.constant 0 : i9001
    scf.if %false {
      %0 = comb.mux %true, %c0_i9001, %c0_i9001 : i9001
    } else {
      %0 = comb.mux %true, %c0_i9001, %c0_i9001 : i9001
    }
  }
  arc.passthrough {
    %c1_i9001 = hw.constant 1 : i9001
    %0 = comb.mux %true, %c1_i9001, %c1_i9001 : i9001
  }
}
```

Reply:
That's a really really nice catch :tada: :100: . I'm now just leaving the insertion point alone, such that all constants simply appear at the top of the clock function in order :+1: